<font size=12> Описание данных и задание SQL</font>

### Задания

* Посчитайте, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитайте количество обзоров и среднюю оценку;
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Базы данных. Описание

<b>Таблица </b>books <br>
 Содержит данные о книгах:<br>
 book_id — идентификатор книги;<br>
 author_id — идентификатор автора;<br>
 title — название книги;<br>
 num_pages — количество страниц;<br>
 publication_date — дата публикации книги;<br>
 publisher_id — идентификатор издателя.<p></p>
<b>Таблица</b> authors <br> 
 Содержит данные об авторах:<br>
 author_id — идентификатор автора;<br>
 author — имя автора. <p></p>
<b>Таблица </b> publishers <br>
 Содержит данные об издательствах:<br>
 publisher_id — идентификатор издательства;<br>
 publisher — название издательства; <p></p>
<b>Таблица </b>ratings <br>
 Содержит данные о пользовательских оценках книг:<br>
 rating_id — идентификатор оценки;<br>
 book_id — идентификатор книги;<br>
 username — имя пользователя, оставившего оценку;<br>
 rating — оценка книги.<p></p>
<b>Таблица</b> reviews <br>
 Содержит данные о пользовательских обзорах на книги:<br>
 review_id — идентификатор обзора;<br>
 book_id — идентификатор книги;<br>
 username — имя пользователя, написавшего обзор;<br>
 text — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
import sqlalchemy as sa
import sqlite3

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas 
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame: 
    '''Открываем соединение, получаем данные из sql, закрываем соединение''' 
    with engine.connect() as con: 
        return pd.read_sql(sql=sa.text(query), con = con) 
    # формируем запрос и выводим данные 
    query = '''SELECT * FROM books LIMIT 5''' 
    get_sql_data(query)

In [3]:
# функция для выполнения запроса к таблице
def query_table(table_name: str) -> pd.DataFrame:
    query = f'''SELECT * FROM {table_name} LIMIT 5'''
    return get_sql_data(query)

In [4]:
query_table('books')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query = '''SELECT COUNT(*) FROM books LIMIT 5'''
print('Количество строк в таблице books:')
get_sql_data(query)

Количество строк в таблице books:


count
0   1000

In [6]:
query_table('authors')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
query = '''SELECT COUNT(*) FROM authors LIMIT 5'''
print('Количество строк в таблице authors:')
get_sql_data(query)

Количество строк в таблице authors:


count
0    636

In [8]:
query_table('publishers')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
query = '''SELECT COUNT(*) FROM publishers LIMIT 5'''
print('Количество строк в таблице publishers:')
get_sql_data(query)

Количество строк в таблице publishers:


count
0    340

In [10]:
query_table('ratings')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
query = '''SELECT COUNT(*) FROM ratings LIMIT 5'''
print('Количество строк в таблице ratings:')
get_sql_data(query)

Количество строк в таблице ratings:


count
0   6456

In [12]:
query_table('reviews')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
query = '''SELECT COUNT(*) FROM reviews LIMIT 5'''
print('Количество строк в таблице reviews:')
get_sql_data(query)

Количество строк в таблице reviews:


count
0   2793

### Сколько книг вышло с 01.01.2000 г

In [14]:
query = '''
SELECT COUNT(*) as num_books
FROM books
WHERE publication_date > '2000-01-01';
'''

result = get_sql_data(query)

num_books = pd.DataFrame(result)
print(num_books)

   num_books
0        819


<font color=blue> С 01 января 2000 года вышло 819 книг. </font>

### Количество обзоров и средняя оценка

Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [15]:
query = '''
SELECT 
    b.book_id,
    b.title,
    COUNT(DISTINCT rw.review_id) AS num_reviews, 
    ROUND(AVG(rt.rating), 2) AS avg_rating
FROM 
    books b
LEFT JOIN 
    reviews rw ON b.book_id = rw.book_id
LEFT JOIN 
    ratings rt ON b.book_id = rt.book_id
GROUP BY 
    b.book_id, b.title;
'''

result = get_sql_data(query)
rating = pd.DataFrame(result)
rating = rating.sort_values(by='num_reviews', ascending=False).reset_index()
# display(rating)

# Суммарное количество отзывов
total_reviews = rating['num_reviews'].sum()

display(rating)

print(f"Суммарное количество отзывов: {total_reviews}")

index  book_id                                              title  \
0      947      948                            Twilight (Twilight  #1)   
1      962      963                                Water for Elephants   
2      732      733                          The Giver (The Giver  #1)   
3      749      750                The Hobbit  or There and Back Again   
4      626      627                                      The Alchemist   
..     ...      ...                                                ...   
995    807      808                            The Natural Way to Draw   
996    190      191  Disney's Beauty and the Beast (A Little Golden...   
997    220      221                          Essential Tales and Poems   
998    386      387                               Leonardo's Notebooks   
999     82       83    Anne Rice's The Vampire Lestat: A Graphic Novel   

     num_reviews  avg_rating  
0              7        3.66  
1              6        3.98  
2              6        3.75  
3              6        4.13  
4              6        3.79  
..           ...         ...  
995            0        3.00  
996            0        4.00  
997            0        4.00  
998            0        4.00  
999            0        3.67  

[1000 rows x 5 columns]

Суммарное количество отзывов: 2793


### Издательство, выпустившее наибольшее число книг с 50+ страницами

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц...

In [16]:
query = '''
SELECT 
    p.publisher, 
    COUNT(b.book_id) AS num_books
FROM 
    books b
JOIN 
    publishers p ON b.publisher_id = p.publisher_id
WHERE 
    b.num_pages > 50
GROUP BY 
    p.publisher
ORDER BY 
    num_books DESC
LIMIT 1;
'''
result = get_sql_data(query)
top_publisher = pd.DataFrame(result)
print('Издательство, которое выпустило наибольшее число книг толще 50 страниц:')
display(top_publisher)

Издательство, которое выпустило наибольшее число книг толще 50 страниц:


publisher  num_books
0  Penguin Books         42

### Автор с самой высокой средней оценкой книг

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [17]:
query = '''
WITH book_ratings AS (
    SELECT 
        book_id,
        AVG(rating) AS avg_rating,
        COUNT(rating_id) AS num_ratings
    FROM 
        ratings
    GROUP BY 
        book_id
    HAVING 
        COUNT(rating_id) >= 50
),
author_ratings AS (
    SELECT 
        b.author_id,
        a.author,
        ROUND(AVG(br.avg_rating), 2) AS author_avg_rating
    FROM 
        book_ratings br
    JOIN 
        books b ON br.book_id = b.book_id
    JOIN 
        authors a ON b.author_id = a.author_id
    GROUP BY 
        b.author_id, a.author
)
SELECT 
    author,
    author_avg_rating
FROM 
    author_ratings
ORDER BY 
    author_avg_rating DESC
LIMIT 1;
'''

result = get_sql_data(query)
top_author = pd.DataFrame(result)
print('Автор с самой высокой средней оценкой книг с 50 и более оценками:')
display(top_author)

Автор с самой высокой средней оценкой книг с 50 и более оценками:


author  author_avg_rating
0  J.K. Rowling/Mary GrandPré               4.28

### Среднее количество обзоров от пользователей

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [18]:
query = '''
WITH users_with_many_ratings AS (
    SELECT 
        username
    FROM 
        ratings
    GROUP BY 
        username
    HAVING 
        COUNT(rating_id) > 48
),
user_reviews AS (
    SELECT 
        rw.username,
        COUNT(rw.review_id) AS num_reviews
    FROM 
        reviews rw
    JOIN 
        users_with_many_ratings uw ON rw.username = uw.username
    GROUP BY 
        rw.username
)
SELECT 
    AVG(num_reviews) AS avg_reviews
FROM 
    user_reviews;
'''

result = get_sql_data(query)
avg_number_of_users = pd.DataFrame(result)
print('Среднее количество обзоров от пользователей, которые поставили больше 48 оценок:')
display(avg_number_of_users)

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок:


avg_reviews
0         24.0

### Выводы

* В базе данных 1000 книг 636 авторов, которых выпустили 340 издательств.
* Было проведено 2793 обзора, а книги были оценены, в общем, 6456 раз.
* Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц - всего 42 книги.
* J.K. Rowling (с иллюстратором Mary GrandPré) является автором с самой высокой средней оценкой книг с 50 и более оценками. Со средней оценкой 4.28.
* Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.